Improve fields.yml generator of modules

kvch opened the pull request:
From now on when a user provides a type hint in an Ingest pipeline, it's added to the generated `fields.yml` instead of guessing.

Closes #7472

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -95,17 +96,18 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
-			Type:     ee[0],
-			Elements: nil,
-		}
+	if 2 != len(ee) && len(ee) != 3 {
```
Could you add a comment here on why we check for these 2 values? Should make reading the code easier.

kvch replied:
Done.

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -95,17 +96,18 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
 func newField(lp string) field {
 	lp = lp[1 : len(lp)-1]
```
This code is fine as long as we use it just as generator on the dev side. But as recently we started to move some code also in the Beats binary we should probably be more careful with validation checks. What happens here in case an empty string is passed to it?

kvch replied:
This function never gets an empty string as a param, because it parses the strings returned by `FindAllString` which does not match for empty lines. But just to be sure, if some tries to use it incorrectly in the future, I added additional error checking.

ruflin replied:
I think the key here is we can't make assumption about future usage of this method as you stated so checks should be in place to make sure the method never panics.

Seems the checks didn't make it in yet? If I run `newField("")` the code would panic?

kvch commented:
I added additional comments and telling names to functions. Furthermore, I renamed the struct attributes to reflect the notions of the Grok processor.

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -95,20 +100,34 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
 func newField(lp string) field {
 	lp = lp[1 : len(lp)-1]
 	ee := strings.Split(lp, ":")
```
What stands `ee` for?

kvch replied:
Well, it's "elements". I don't really have a good name for elements of `SYNTAX:SEMANTIC:TYPE` split at `:`s. I can rename it to `elements`, but I don't think it would help. :(

ruflin replied:
I quite like `elements`, sounds better to me then `ee`

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
-		}
+func newField(pattern string) field {
+	if len(pattern) <= 2 {
+		return field{}
```
Should we log errors instead of return empty fields? This would make debugging much easier in the future.

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
+
+	elements := strings.Split(pattern, ":")
+	if !isValidFormat(elements) {
+		return field{}
```
Return error?

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
+	if len(pattern) <= 2 {
+		return field{}
+	}
+	pattern = pattern[1 : len(pattern)-1]
```
Alternative here would `strings.Trim("pattern", "{}")`

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
+}
+
+// the last element is the type hint
+func containsType(ee []string) bool {
```
Should this be `containsHint`? I understand the the hint is for the type but this checks if it contains hint?

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
+		return field{}
+	}
+
+	hint := ""
```
What about having a function `getType(elements)` that does the below? Would also make it easier to test I think.

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -92,23 +97,41 @@ func newFieldYml(name, typeName string, noDoc bool) *fieldYml {
-		Elements: e,
+		Syntax:           elements[typeIdx],
+		SemanticElements: strings.Split(elements[elementsIdx], "."),
+		Type:             hint,
```
The variable should be called `type` probably, see the above.

ruflin commented on filebeat/scripts/generator/fields/main.go:
```diff
@@ -33,6 +33,10 @@ import (
 	pipelinePath  = "%s/module/%s/%s/ingest/pipeline.json"
 	fieldsYmlPath = "%s/module/%s/%s/_meta/fields.yml"
+
+	typeIdx     = 0
```
Isn't that the syntax index?

kvch commented:
I would rather not address all your review comments in this PR as I already have another one specifically for refactoring and making this script ready to be used as a subcommand. I intended this PR to be a quick fix orthogonal to refactoring, so I can rebase my other PR on top of it with minimal conflicts.

So what do you think if we merge this PR and I do the refactoring you requested in https://github.com/elastic/beats/pull/7506?

ruflin commented:
Merged as we discussed that refactoring of the comments will happen in https://github.com/elastic/beats/pull/7506 or a separate PR.
